src/circuit.rs

```diff
@@ -37,13 +39,15 @@ impl Circuit {
             gates: vec![],
             edges: vec![],
             ends: vec![None; n_elements],
+            gate_elements: vec![],
         }
     }
 
     pub fn add(&mut self, gate: Gate, elements: Vec<usize>) -> usize {
         self.gates.push(gate);
         // retrieve gate ID
         let gid = self.gates.len() - 1;
+        self.gate_elements.push(elements.len());
```

Comment:
Since we're not yet aiming for performances, do we really need to push the gate elements?

This should be "redundant" with the information in `edges`, i.e. you could implement `self.gate_elements(gid)` just by filtering the edges, looking at the second member, for the nodes with that given `gid`.
I.e.:
```rs
fn gate_elements(&self, gid: usize) -> usize {
    self.edges.filter(|e| e.1.gid == gid).count()
}
```

Reply:
It's not just to make a weird exercise about how much we could complicate our lives, but rather "the opposite": since it's already not performant, let's keep it simple, with fewer moving parts around.

Whenever we'll make an effort for performances, we'll do it consistently, assessing where we could gain the most benefits.

Reply:
624835be7eeb0b54d0491507819e74e60c929866